rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{Error, KernelResult};
+use crate::{bindings, c_types, CStr};
+use alloc::boxed::Box;
+use core::marker::PhantomPinned;
+use core::mem::transmute;
+use core::pin::Pin;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_get_drvdata(
+        pdev: *const bindings::platform_device,
+    ) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_set_drvdata(
+        pdev: *mut bindings::platform_device,
+        data: *mut c_types::c_void,
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = Box::try_new(drv_data)?;
+        let drv_data = Box::into_raw(drv_data) as *mut c_types::c_void;
+        Ok(drv_data) as KernelResult<_>
+    };
+    // TODO don't we need to pin this?
+    let ptr = match f() {
+        Ok(ptr) => ptr,
+        Err(e) => return e.to_kernel_errno(),
+    };
+    rust_helper_platform_set_drvdata(pdev, ptr);
+    0
+}
+
+fn new_of_device_id(compatible: &CStr<'static>) -> KernelResult<bindings::of_device_id> {
+    // TODO:
+    // - fail at build time if compatible CStr doesn't fit.
+    // - can we do this safely without transmute?
+    let mut buf = [0_u8; 128];
+    if compatible.len() > buf.len() {
+        return Err(Error::EINVAL);
+    }
+    // PANIC: this will never panic: `compatible` is not longer than `buf`.
+    buf[..compatible.len()].copy_from_slice(compatible.as_bytes());
```

Comment:
This could probably be
```rust
let dest = buf.get_mut(..compatible.len()).ok_or(Error::EINVAL)?;
    
dest.copy_from_slice(compatible.to_bytes());
```
which removes the `PANIC` comment / the duplicated bounds check.

Reply:
Nice! Can you (or anyone) think of a way to perform the check at build time?

Reply:
I don't think there is a way with the current function signature as the length of the `&CStr` is only decided at runtime from the view of the function. It may be possible to do this with a macro but I don't know if the additional complexity that would introduce is worth the one bounds check. I'll see if I can write something that's not too complex if I find the time.

Reply:
I thought about this a bit more and I guess it would be possible take the [i8; 128] (possibly wrapped in a newtype) as an argument instead of a `&CStr` (which needs to be propagated to `new_pinned` etc.) and then use a macro to create that from a string literal at compile time when calling these function (instead of `cstr!`). That way the ergonomics impact would not be that big but I still don't know if that is worth it and if the compiler can sufficiently optimise the array copies to be faster than a single bounds check.